src/Test.Utilities/CodeFixRunner.cs

```diff
@@ -0,0 +1,299 @@
+                        newCompilerDiagnostics.Select(d => d.ToString()).Join("\r\n"),
+                        project.Documents.Select(doc => doc.GetSyntaxRootAsync().Result.ToFullString()).Join("\r\n")));
+            }
+        }
+
+        private static IEnumerable<Diagnostic> GetNewDiagnostics(IEnumerable<Diagnostic> diagnostics, IEnumerable<Diagnostic> newDiagnostics)
+        {
+            Diagnostic[] oldArray = diagnostics.OrderBy(d => d.Location.SourceSpan.Start).ToArray();
+            Diagnostic[] newArray = newDiagnostics.OrderBy(d => d.Location.SourceSpan.Start).ToArray();
+
+            int oldIndex = 0;
+            int newIndex = 0;
+
+            while (newIndex < newArray.Length)
+            {
+                if (oldIndex < oldArray.Length && oldArray[oldIndex].Id == newArray[newIndex].Id)
+                {
+                    ++oldIndex;
+                    ++newIndex;
+                }
+                else
+                {
+                    yield return newArray[newIndex++];
+                }
+            }
+        }
+
+        protected Diagnostic[] GetSortedDiagnostics(Compilation compilation, IEnumerable<TestAdditionalDocument> additionalFiles = null)
+        {
+            if (_analyzerOpt == null)
+            {
+                return Array.Empty<Diagnostic>();
+            }
+
+            var analyzerOptions = additionalFiles != null ? new AnalyzerOptions(additionalFiles.ToImmutableArray<AdditionalText>()) : null;
+            DiagnosticBag diagnostics = DiagnosticBag.GetInstance();
+            compilation = EnableAnalyzer(_analyzerOpt, compilation);
+
+            return compilation.GetAnalyzerDiagnostics(new[] { _analyzerOpt }, _validationMode, analyzerOptions).OrderBy(d => d.Location.SourceSpan.Start).ToArray();
+        }
+
+        private static Compilation EnableAnalyzer(DiagnosticAnalyzer analyzer, Compilation compilation)
+        {
+            return compilation.WithOptions(
+                compilation
+                    .Options
+                    .WithSpecificDiagnosticOptions(
+                        analyzer
+                            .SupportedDiagnostics
+                            .Select(x => KeyValuePair.Create(x.Id, ReportDiagnostic.Default))
+                            .ToImmutableDictionaryOrEmpty()));
+        }
+
+        private class FixAllDiagnosticProvider : FixAllContext.DiagnosticProvider
+        {
+            private IEnumerable<TestAdditionalDocument> _additionalFiles;
+            private readonly CodeFixRunner _runner;
+
+            public FixAllDiagnosticProvider(
+                CodeFixRunner runner,
+                IEnumerable<TestAdditionalDocument> additionalFiles)
+            {
+                _additionalFiles = additionalFiles;
+                _runner = runner;
+            }
+
+            public override async Task<IEnumerable<Diagnostic>> GetDocumentDiagnosticsAsync(Document document, CancellationToken cancellationToken)
+            {
```

Comment:
Hopefully this will be simplified once the `FixAllDiagnosticProvider` is directly provided the `fixableDiagnostics` in the constructor. You can also consider implementing `GetAllDiagnosticsAsync` and `GetProjectDiagnosticsAsync`